### ISSUE TITLE
[Request] Server side routing

### ISSUE THREAD

**Issue by tomswinkels** [stefanak-michal/Bolt#18]
We are running Neo4j in cluster, we want use the Neo4j server side routing.

Maybe this is a good feature for this package?

I cant find any documentation how the server side routing works, maybe you have examples?

**Comment by tomswinkels** [comment]
I am trying to make some wrapper for this.

But when i added more connections in a foreach a get the error `Not initialized socket` is there some option to reset/kill the connection?

**Comment by stefanak-michal** [comment]
- The BEGIN message, defines the sub-field routing::Dictionary(address::String) to indicate if server side routing should be performed and can include routing context data. ([source](https://7687.org/bolt/bolt-protocol-message-specification-4.html#deltas-1))

That means you can use it with my library, but I was missing extra argument on few places. It's described here: https://7687.org/bolt/bolt-protocol-message-specification-4.html#request-message---41---hello

It will look like this:
```php
$bolt = new \Bolt\Bolt();
if (!$bolt->init('MyClient/1.0', $user, $password, [ 'routing' => ['address' => 'x.example.com:9001'] ])) {
    throw new Exception('Wrong login');
}
```

I've added missing extra parameter in this branch: https://github.com/stefanak-michal/Bolt/tree/extra_parameters

https://github.com/stefanak-michal/Bolt/commit/8dcbbb94cbb3270751c1a5f5f0d8deb28169e371

_I don't have cluster instance which means I have no way to test it._

**Comment by stefanak-michal** [comment]
> I am trying to make some wrapper for this.
> 
> But when i added more connections in a foreach a get the error `Not initialized socket` is there some option to reset/kill the connection?

It can be because Socket.php holds socket connection in static variable. I'll change that.

You can kill connection with destroying Bolt instance. It contains destructor which is closing socket.

https://github.com/stefanak-michal/Bolt/blob/master/Bolt.php#L366

**Comment by tomswinkels** [comment]
Nice.

But in the first example i must set a connection, if the first connection server is down than it does not work.

**Comment by stefanak-michal** [comment]
Routing to bolt protocol was added in version 4.1

**Comment by stefanak-michal** [comment]
removed Socket.php static access

https://github.com/stefanak-michal/Bolt/tree/issue-18-socket_non_static

**Comment by tomswinkels** [comment]
Nice, that works! 

I was writing a wrapper for version 3. But we have change plans to first upgrade to neo4j 4.1 and than fix routing!

`        
        $writeServers = [];
        $readServers = [];

        $neo4jClusters = [
            '10.10.68.13', //Offline server
            '10.10.68.14', //Server1
            '10.10.68.15', //Server2
            '10.10.68.47' //Server3
        ];

        if($neo4jClusters) {
            foreach($neo4jClusters as $neo4jCluster) {

                try {
                    $bolt = new \Bolt\Bolt($neo4jCluster, $port, 1);
                    $bolt->setProtocolVersions(3);
                    $bolt->init('MyClient/1.0', $username, $password);

                    //Execute query
                    $bolt->run('CALL dbms.cluster.routing.getServers()');
                    //Pull records from last query
                    $rows = $bolt->pull();

                    if(isset($rows[0][1])) {

                        foreach($rows[0][1] as $server) {

                            if($server['role'] == 'WRITE') {
                                $writeServers = $server['addresses'];
                            } elseif($server['role'] == 'READ') {
                                $readServers = $server['addresses'];
                            }
                        }
                    }

                } catch (\Exception $e) {

                    dump('Failed to connect');

                    //Try some other server
                    continue;
                }
            }
        }
`

**Comment by tomswinkels** [comment]
I have read that you can also set a cluster on one server 👍 
https://neo4j.com/docs/operations-manual/current/tutorial/local-causal-cluster/

**Comment by stefanak-michal** [comment]
I created cluster on my PC and tested with Neo4j 4.1.3.

![obrázok_2020-10-16_215556](https://user-images.githubusercontent.com/5502917/96303241-5eb39480-0ffa-11eb-9380-353d4fbaa358.png)

It acts how it should be. I checked :sysinfo (browser interface at :7474) and discovered Leader of my cluster. Therefore I connected to specified port and it works.
```php
$bolt = new \Bolt\Bolt('localhost',7688);
```
When I tried to change to Follower port I got error: _No write operations are allowed directly on this database. Writes must pass through the leader. The role of this server is: FOLLOWER (Neo.ClientError.Cluster.NotALeader)_

Which is correct.

Also is possible discover cluster Leader and Follower with query in this newer Neo4j version
```
CALL dbms.routing.getRoutingTable({})
```

Now I have to try your solution for older version.
